RabbitMQ: Support for external authentication with SASL

abossard opened the pull request:
# Description

I added a parameter to support external sasl authentication based on the Client Certificate. 

This is based on the EXTERNAL mechanism described here:
https://www.rabbitmq.com/access-control.html#mechanisms

And an example RabbitMQ Docker Image that uses the configuration is located here:
https://github.com/GFlisch/RabbitmqS

This means it will take 
## Issue reference
#2514 

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2514 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#3184

berndverst commented:
Please make sure you sign off on all your commits first. See https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work

berndverst commented on pubsub/rabbitmq/metadata.go:
```diff
@@ -223,6 +226,15 @@ func createMetadata(pubSubMetadata pubsub.Metadata, log logger.Logger) (*metadat
 	}
 
+	if val, found := pubSubMetadata.Properties[metadataSaslExternal]; found && val != "" {
+		if boolVal, err := strconv.ParseBool(val); err == nil {
```
Instead of `ParseBool` please use our `utils.IsTruthy` function https://github.com/dapr/components-contrib/blob/master/internal/utils/utils.go#L26

abossard replied:
Thanks, I replaced it. ✌️

berndverst reviewed (changes requested):
Please add an E2E test here:
https://github.com/dapr/components-contrib/tree/master/tests/certification/pubsub/rabbitmq

You will need to create a new `component.yaml` file to explictly test the `salsExternal` property you are introducing.

The test should spin up a RabbitMQ docker container with the required settings and this `salsExternal` property set in such a way that you can verify it works end to end.

berndverst commented on pubsub/rabbitmq/metadata.go:
```diff
@@ -15,6 +15,7 @@ package rabbitmq
 
 import (
 	"fmt"
+	"github.com/dapr/components-contrib/internal/utils"
```
for the linter to pass you will need to remove dapr imports into a separate block below standard library things.

```
stdlib stuff
<blank>
dapr imports
```

or if you have third party imports also

```
stdlib stuff
<blank>
third party imports
<blank>
dapr imports
```

abossard replied:
Thanks, I sorted it now 👌

abossard commented:
> Please add an E2E test here: https://github.com/dapr/components-contrib/tree/master/tests/certification/pubsub/rabbitmq
> 
> You will need to create a new `component.yaml` file to explictly test the `salsExternal` property you are introducing.
> 
> The test should spin up a RabbitMQ docker container with the required settings and this `salsExternal` property set in such a way that you can verify it works end to end.

Should I add another container to the existing compose.yaml or make a entirely new folder? 
For the component, I assume I just make a folder `mtls_external` and put the special component in there?

berndverst commented:
> > Please add an E2E test here: https://github.com/dapr/components-contrib/tree/master/tests/certification/pubsub/rabbitmq
> > You will need to create a new `component.yaml` file to explictly test the `salsExternal` property you are introducing.
> > The test should spin up a RabbitMQ docker container with the required settings and this `salsExternal` property set in such a way that you can verify it works end to end.
> 
> Should I add another container to the existing compose.yaml or make a entirely new folder? For the component, I assume I just make a folder `mtls_external` and put the special component in there?

Yes make a new folder and put your component yaml there. Then in the test code make sure you load the correct component yaml files of course :)

abossard commented:
Hi @berndverst I added an E2E test. I added the additional Rabbit MQ instance to the existing docker-compose.yaml. Is that alright, or should it be in it's separate docker-compose.yaml?

I reduced the test to the simple direct `test` without the network connectivity tests, since I wasn't able to successfully run them. But I think to check if the mTLS with External Auth is working it should be fine.

The certificates supplied to the tests run out next year.

abossard commented:
> Please create a a separate yaml and a separate folder for that yaml. Basically everything this test needs in terms of configuration should be self contained in another folder. Then make sure in your test your test run it starts up by loading this new yaml file.

Got it, thanks. Added it: https://github.com/dapr/components-contrib/pull/2515/commits/35acca5f7b7436cfd24ff63b27e73325dcb21059

abossard commented:
Hi @berndverst it should be fine now, right?

berndverst commented on pubsub/rabbitmq/metadata.go:
```diff
@@ -19,6 +19,8 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/dapr/components-contrib/internal/utils"
```
this will probably fail the linter because the dapr libraries must be the last import block

berndverst commented:
```
=== Failed
=== FAIL: . TestRabbitMQExtAuth/rabbitmq_mtls_external_authentication (1.30s)
    flow.go:220: Running step: rabbitmqcertification
    dockercompose.go:57: build path /home/runner/work/components-contrib/components-contrib/tests/certification/pubsub/rabbitmq/mtls_sasl_external/mtls_sasl_external/docker_sasl_external either does not exist, is not accessible, or is not a valid URL.
        
    flow.go:228: Completed step: rabbitmqcertification
    flow.go:230: Fatal error in step rabbitmqcertification: exit status 1
    dockercompose.go:72: build path /home/runner/work/components-contrib/components-contrib/tests/certification/pubsub/rabbitmq/mtls_sasl_external/mtls_sasl_external/docker_sasl_external either does not exist, is not accessible, or is not a valid URL.
    ```

berndverst reviewed (changes requested):
Certification test is failing due to the docker file path

berndverst commented on tests/certification/pubsub/rabbitmq/mtls_sasl_external/docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+services:
+  rabbitmq_external_sasl:
+    image: rabbitmq_external_sasl
+    build: mtls_sasl_external/docker_sasl_external
```
Is this perhaps a relative path that needs to be updated?

abossard replied:
I assume so much, yes. Thanks. I don't think I've ever pruned all images. I did it now, and it works with ways, but maybe docker desktop compose might have a different path lookup scheme....
